docs/security-considerations.rst

```diff
@@ -344,3 +346,118 @@ Note that formal verification itself can only help you understand the
 difference between what you did (the specification) and how you did it
 (the actual implementation). You still need to check whether the specification
 is what you wanted and that you did not miss any unintended effects of it.
+
+Solidity implements a formal verification approach based on SMT solving.  The
+SMTChecker module automatically tries to prove that the code satisfies the
+specification given by ``require/assert`` statements, that is, it considers
+``require`` statements as assumptions and tries to prove that the conditions
+inside ``assert`` statements are always true.  If an assertion failure is
+found, a counterexample is given to the user, showing how the assertion can be
+violated.
+
+The SMTChecker also checks automatically for underflow/overflow, trivial
+conditions and unreachable code.
+It is currently an experimental feature, therefore in order to use it you need
+to enable it via :ref:`a pragma directive<smt_checker>`.
+
+The SMTChecker traverses the Solidity AST creating and collecting program constraints.
+When it encounters a verification target, an SMT solver is invoked to determine the outcome.
+
+For more details on how the SMT encoding works internally, see
+`our paper <https://github.com/leonardoalt/text/blob/master/solidity_isola_2018/main.pdf>`_.
+
+Abstraction and false positives
+===============================
+
+The SMTChecker implements abstractions in an incomplete and sound way: If a bug
+is reported, it might be a false positive introduced by abstractions (while
+erasing knowledge or using a non-precise type), but if it determines that a
+verification target is safe, it is indeed safe (there are no false negatives).
+
+The SMT encoding tries to be as precise as possible, mapping Solidity types and
+expressions to their closest SMT representation, as shown in the table below.
+
++-----------------------+--------------+-----------------------------+
+|Solidity type          |SMT sort      |Theories (quantifier-free)   |
++=======================+==============+=============================+
+|Boolean                |Bool          |Bool                         |
++-----------------------+--------------+-----------------------------+
+|intN, uintN, address   |Integer       |LIA, NIA                     |
+|bytesN, enum           |              |                             |
++-----------------------+--------------+-----------------------------+
+|array, mapping         |Array         |Arrays                       |
++-----------------------+--------------+-----------------------------+
+|other types            |Integer       |LIA                          |
++-----------------------+--------------+-----------------------------+
+
+Types that are not yet supported are abstracted by a single 256-bit unsigned integer,
+where their unsupported operations are ignored.
+
+Function calls are inlined when possible, otherwise they are abstracted by an
+uninterpreted function over the arguments.
```

Comment:
Oh that's interesting! Does that mean it is an actual function? Because we cannot guarantee that due to state in storage.

Reply:
Yep, it's an actual function. What can't we guarantee? If it's an external function storage knowledge is erased, not sure if that's what you meant.

Reply:
Will we always inline here? Might there be a reason not to inline?
```
contract c {
  uint x;
  function f() internal returns (uint) { return ++x; }
  function g() public {
    assert(f() == f());
  }
}
```
Also external:
```
contract C {
  uint x;
  function f() external returns (uint) { return ++x; }
  function g() public {
    assert(this.f() == this.f()); // I guess this is modelled as a generic external call
  }
}
```

Reply:
The first case would inline.
I'd say we should inline the second as well.

I guess I should rephrase the text, saying that we abstract functions we do not have access to the code.

Reply:
Ok, but does the SMT checker complain in both cases? We should probably add these as test cases.

Reply:
One reason not to always inline would be short-circuiting

Reply:
It complains correctly in the first case, and complains in the second due to lack of support to `this.`